File: client-ts/src/delta.ts
// ── Delta op types ─────────────────────────────────────────────────────────────

export type DeltaOp =
  | { op: "ADD";    path: string; value: unknown }
  | { op: "UPDATE"; path: string; value: unknown }
  | { op: "REMOVE"; path: string };

// ── Deep clone helper ─────────────────────────────────────────────────────────

/** Deep clone that works on Node 16 and older browsers lacking structuredClone. */
function deepClone<T>(value: T): T {
  if (typeof globalThis.structuredClone === "function") {
    return globalThis.structuredClone(value);
  }
  return JSON.parse(JSON.stringify(value)) as T;
}

// ── Path parsing ───────────────────────────────────────────────────────────────

/**
 * Parse a JSON-pointer-style path into its segments.
 *
 * "/"         → []           (root)
 * "/x"        → ["x"]
 * "/x/y/z"   → ["x", "y", "z"]
 * "/-"        → ["-"]        (RFC 6901 array append token)
 */
function parsePath(path: string): string[] {
  if (path === "/") return [];
  return path.slice(1).split("/");
}

// ── Known op types ────────────────────────────────────────────────────────────

const KNOWN_OPS = new Set(["ADD", "UPDATE", "REMOVE"]);

// ── Op application ─────────────────────────────────────────────────────────────

/**
 * Apply a sequence of delta ops to `state` and return the new state value.
 * Does not mutate the input — clones before modification.
 *
 * Throws on unknown op types.
 */
export function applyOps(state: unknown, ops: DeltaOp[]): unknown {
  for (const op of ops) {
    state = applyOp(state, op);
  }
  return state;
}

function applyOp(state: unknown, op: DeltaOp): unknown {
  if (!KNOWN_OPS.has(op.op)) {
    throw new Error(`[SODP] unknown delta op type: "${op.op}". Expected one of: ADD, UPDATE, REMOVE`);
  }

  const parts = parsePath(op.path);

  // Root-level operation.
  if (parts.length === 0) {
    if (op.op === "REMOVE") return null;
    return (op as { value: unknown }).value;
  }

  // Clone so callers keep immutability guarantees.
  const root = (
    typeof state === "object" && state !== null
      ? deepClone(state)
      : {}
  ) as Record<string, unknown>;

  // Navigate to the parent node, creating intermediate objects as needed.
  let node: unknown = root;
  for (let i = 0; i < parts.length - 1; i++) {
    const key = parts[i];
    const parent = node as Record<string, unknown>;
    if (typeof parent[key] !== "object" || parent[key] === null) {
      parent[key] = {};
    }
    node = parent[key];
  }

  const last = parts[parts.length - 1];

  // RFC 6901: "-" references the element past the end of an array (append).
  if (last === "-" && Array.isArray(node)) {
    switch (op.op) {
      case "ADD":
      case "UPDATE":
        (node as unknown[]).push((op as { value: unknown }).value);
        break;
      case "REMOVE":
        // Remove last element when targeting "-" on a REMOVE
        (node as unknown[]).pop();
        break;
    }
    return root;
  }

  const parent = node as Record<string, unknown>;
  switch (op.op) {
    case "ADD":
    case "UPDATE":
      // If parent is an array and key is numeric, set by index.
      if (Array.isArray(parent)) {
        const idx = Number(last);
        if (!Number.isNaN(idx)) {
          (parent as unknown[])[idx] = (op as { value: unknown }).value;
          break;
        }
      }
      parent[last] = (op as { value: unknown }).value;
      break;
    case "REMOVE":
      if (Array.isArray(parent)) {
        const idx = Number(last);
        if (!Number.isNaN(idx)) {
          (parent as unknown[]).splice(idx, 1);
          break;
        }
      }
      delete parent[last];
      break;
  }

  return root;
}

File: sodp-py/src/sodp/delta.py
"""Apply SODP delta ops to an in-memory state value.

Each op is a dict with keys ``op``, ``path``, and (for ADD/UPDATE) ``value``.
Paths follow JSON Pointer (RFC 6901):

* ``"/"``   — the root value itself
* ``"/x"``  — top-level field ``x`` on an object
* ``"/x/y"`` — nested field
* ``"/-"``  — the element past the end of an array (append)
* ``"/3"``  — index ``3`` into an array
"""

from __future__ import annotations

import copy
from typing import Any

_KNOWN_OPS = frozenset({"ADD", "UPDATE", "REMOVE"})


def apply_ops(state: Any, ops: list[dict]) -> Any:
    """Return the new state after applying *ops* in order. Does not mutate *state*.

    Raises :class:`ValueError` if any op has an unknown ``op`` type (anything
    other than ``"ADD"``, ``"UPDATE"``, or ``"REMOVE"``). This is deliberate —
    silently ignoring unknown ops causes the client's cached state to diverge
    from the server, which is far worse than a loud failure.
    """
    for op in ops:
        state = _apply_one(state, op)
    return state


def _apply_one(state: Any, op: dict) -> Any:
    kind: str = op.get("op", "")
    if kind not in _KNOWN_OPS:
        raise ValueError(
            f"[SODP] unknown delta op type: {kind!r}. "
            f"Expected one of: ADD, UPDATE, REMOVE"
        )

    path: str = op.get("path", "/")
    parts = _parse_path(path)

    # Root operation.
    if not parts:
        return None if kind == "REMOVE" else op.get("value")

    # Deep-clone the container so callers keep immutability guarantees.
    # Non-container state at a non-root path is coerced to an empty dict.
    if isinstance(state, (dict, list)):
        root: Any = copy.deepcopy(state)
    else:
        root = {}

    # Walk to the parent node, materialising intermediate dicts as needed.
    node: Any = root
    for key in parts[:-1]:
        if isinstance(node, list):
            idx = _as_index(key)
            if idx is None or not (0 <= idx < len(node)):
                # Non-indexable segment on a list — nothing we can sensibly do.
                return root
            node = node[idx]
        else:
            child = node.get(key) if isinstance(node, dict) else None
            if not isinstance(child, (dict, list)):
                child = {}
                if isinstance(node, dict):
                    node[key] = child
            node = child

    last = parts[-1]

    # RFC 6901 "-" append on an array.
    if last == "-" and isinstance(node, list):
        if kind in ("ADD", "UPDATE"):
            node.append(op.get("value"))
        elif kind == "REMOVE" and node:
            node.pop()
        return root

    if isinstance(node, list):
        idx = _as_index(last)
        if idx is None:
            return root
        if kind in ("ADD", "UPDATE"):
            if 0 <= idx < len(node):
                node[idx] = op.get("value")
            elif idx == len(node):
                node.append(op.get("value"))
        elif kind == "REMOVE" and 0 <= idx < len(node):
            node.pop(idx)
        return root

    # Dict branch.
    if not isinstance(node, dict):
        return root
    if kind in ("ADD", "UPDATE"):
        node[last] = op.get("value")
    elif kind == "REMOVE":
        node.pop(last, None)

    return root


def _as_index(seg: str) -> int | None:
    """Parse *seg* as a non-negative array index, or return ``None``."""
    if not seg or not seg.isdigit():
        return None
    return int(seg)


def _parse_path(path: str) -> list[str]:
    """``"/"`` → ``[]``, ``"/x/y"`` → ``["x", "y"]``, ``"/-"`` → ``["-"]``."""
    if path == "/":
        return []
    return path[1:].split("/")
